src/ExpressionInterface.php

```diff
@@ -2,21 +2,31 @@
 
 namespace Dhii\Expression;
 
-use Dhii\Evaluable\EvaluableInterface;
+use Dhii\Util\String\StringableInterface as Stringable;
+use Traversable;
 
 /**
- * An expression is an evaluable construct that is made up of 1 or more evaluable terms.
+ * Something that represents an expression.
  *
  * @since 0.1
  */
-interface ExpressionInterface extends EvaluableInterface
+interface ExpressionInterface
 {
     /**
-     * Gets the expression terms.
+     * Retrieves the expression type.
+     *
+     * @since [*next-version*]
+     *
+     * @return string|Stringable The expression type code.
+     */
+    public function getType();
+
+    /**
+     * Retrieves the terms for this expression.
      *
      * @since 0.1
      *
-     * @return EvaluableInterface[] An array of evaluable instances.
+     * @return string[]|Stringable[]|ExpressionInterface[]|Traversable A list of term strings or sub-expressions.
```

Comment:
How come terms can be strings? Wouldn't it make more sense to require them to be expressions?

Reply:
I considered having all nodes in the hierarchy be restricted to expressions. I took this approach so as to be able to "formally" expose leaf nodes in the hierarchy, because if all terms are expressions, then no leaf values can be obtained.

Say if you had `(1 * 2) + 3`, you'd have an addition expression with 2 terms: a multiplication expression and `"3"`.

I assume that what you're asking/suggesting is implying that the second term would not be a `"3"` string, but a `LiteralIntegerExpression` or something. If that were the case, no uniform expression crawling algorithm would be able to obtain the `"3"`.

Reply:
Yes, that's what I was implying. And you are right. But I still don't like the `string` here - or `Stringable`, for that matter. Can we go around this somehow? Perhaps using `ValueAwareInterface`?

Reply:
That's actually a very good idea. I like it.

Reply:
Would the doc be correct if it stated the parameter type to be the `EvaluableInterface[]|ExpressionInterface[]|Traversable` since they can be mixed?

Reply:
If this is better in your opinion, then I think you should do this.

Reply:
I understand the question now. I don't know of a better way. But perhaps, this complication means that we are approaching the problem in the wrong way.

Reply:
Perhaps. Maybe we can think of a common interface for expressions and terms and use that here?

Or are we, once again, overthinking this by thinking that doc limitations indicate design problem? :P If it were anyone else, they would just type `array` and be over with it. haha

Reply:
Fortunately or not, we are not anyone else. :P

A common interface means that the items would have to be both things.

Reply:
Not necessarily. A common interface `A`, extended by both `B` and `C` doesn't make `B`s the same as `C`s and vice-versa.

What about this? We introduce `TermInterface` with `getType()`, the same `getType()` that is currently in `ExpressionInterface`.

`ExpressionInterface` then naturally extends that and adds `getTerms()`.

A `LiteralTermInterface` can then be introduced that extends `TermInterface` and `ValueAwareInterface`.

Reply:
The more I think about this, the more it makes sense. An expression is something that expresses value, be it a literal value or an expansion of operations. At their core, they are the same. But in this package, we are not concerned with evaluation.

When it comes to recursive hierarchy, I think a common type is a must. Perhaps the naming can be improved, but I feel like `ExpressionInterface` ought to be reserved for composite things, i.e. a term that has child terms.

It also sounds very familiar. I feel like this is something similar to what I had prototyped a long time ago for expressions.

Reply:
> Not necessarily. A common interface A, extended by both B and C doesn't make Bs the same as Cs and vice-versa.

For some reason I thought that you mean `interface A extends B, C`. My bad.

With regard to the last comment, yes, that makes sense. I remember that in the beginning of the whole expression direction I was struggling to decide what is best: expression extends term, or term extends expression? At the time, thee were no cases that could suggest to me that one of the answers is better. But now, it appears that `ExpressionInterface extends TermInterface` is the way to go.

Please go ahead and try it out, if you agree with me.

Reply:
Yes, I recall a brief discussion at one point where we talked about which way would be better.

I'm a bit surprised at how simple and elegant this solution sounds, and why this wasn't the initial direction, tbh 🤣